cachito/web/api_v1.py

```diff
@@ -744,68 +744,134 @@ def get_request_logs(request_id):
-                f"not complete or stale."
-            )
         request_ids.remove(request.id)
 
     if request_ids:
         nonexistent_ids = ",".join(map(str, request_ids))
         raise BadRequest(f"Cannot find request(s) {nonexistent_ids}.")
 
+    return requests
+
+
+def _check_requests_state(requests: List[Request], valid_states: Set[str]) -> None:
+    error_msg = ""
+
+    request: Request
+    for request in requests:
+        if request.state.state_name not in valid_states:
+            error_msg += (
+                f"Request {request.id} is in state {request.state.state_name}, "
+                f"not in {valid_states}.\n"
+            )
+
+    if error_msg:
+        raise BadRequest(error_msg)
+
+
+@api_v1.route("/content-manifest", methods=["GET"])
+def get_content_manifest_by_requests():
+    """
+    Retrieve the content manifest associated with the given requests.
+
+    :return: a Flask JSON response
+    :rtype: flask.Response
+    :raise BadRequest: if any of the given request is not in the "complete" or
+        "stale" state, If any of the given request cannot be found.
+    """
+    arg = flask.request.args.get("requests")
+    if not arg:
+        return flask.jsonify(BASE_ICM)
+
+    request_ids = _get_valid_request_ids(arg)
+    requests = _get_all_requests(request_ids)
+
+    valid_states = set([RequestStateMapping.complete.name])
+    _check_requests_state(requests, valid_states)
+
+    request: Request
     assembled_icm = deepcopy(BASE_ICM)
     for request in requests:
         manifest = request.content_manifest.to_json()
         assembled_icm["image_contents"].extend(manifest["image_contents"])
     if len(requests) > 1:
         deep_sort_icm(assembled_icm)
-    return send_content_manifest_back(assembled_icm)
+    return send_json_file_back(assembled_icm)
+
+
+@api_v1.route("/sbom", methods=["GET"])
+def get_sbom_by_requests() -> flask.Response:
+    """
+    Retrieve the content manifest sbom associated with the given requests.
+
+    :return: a Flask JSON response
+    :rtype: flask.Response
+    :raise BadRequest: if any of the given request is not in the "complete" or
+        "stale" state, If any of the given request cannot be found.
+    """
+    arg = flask.request.args.get("requests")
+    if not arg:
+        return flask.jsonify(BASE_SBOM)
+
+    request_ids = _get_valid_request_ids(arg)
+    requests = _get_all_requests(request_ids)
+
+    valid_states = set([RequestStateMapping.complete.name])
+    _check_requests_state(requests, valid_states)
+
+    request: Request
+    assembled_sbom = deepcopy(BASE_SBOM)
+
+    all_components = []
+    for request in requests:
+        sbom_components = request.content_manifest.sbom_components_list()
+        all_components.extend(sbom_components)
+
+    unique_components = []
+    for component in all_components:
+        if component not in unique_components:
+            unique_components.append(component)
```

Comment:
ah, I was hoping to avoid n<sup>2</sup> complexity by de-duplicating once

It could be done like this I think
```python
all_components.sort(key=lambda c: (c["purl"], c["name"], c["version"])
for component in all_components:
    if not unique_components or component != unique_components[-1]:
        unique_components.append(component)
```
This should have better time complexity, plus we won't have to call `deep_sort_icm` on the SBOM anymore

Reply:
ok changed